File: backend-app/crud/src/main/java/it/uniba/crud/service/CaregiverService.java
package it.uniba.crud.service;

import java.util.List;
import java.util.Optional;

import it.uniba.crud.dto.CaregiverDTO;
import it.uniba.crud.entity.Caregiver;

public interface CaregiverService {
    /**
     * Recupera tutti i caregiver
     * @return Lista di caregiver
     */
    List<Caregiver> fetchAllCaregiver();

    /**
     * Recupera un caregiver per ID
     * @param id ID del caregiver
     * @return Caregiver trovato
     */
    Caregiver fetchById(Long id);

    /**
     * Recupera un caregiver per ID in modo sicuro
     * @param id ID del caregiver
     * @return Optional del caregiver
     */
    Optional<Caregiver> fetchByIdSafe(Long id);

    /**
     * Crea un nuovo caregiver
     * @param caregiver DTO del caregiver da creare
     * @return Caregiver creato
     */
    Caregiver createCaregiver(CaregiverDTO caregiver);

    /**
     * Aggiorna un caregiver esistente
     * @param id ID del caregiver da aggiornare
     * @param caregiver DTO con i nuovi dati
     * @return Caregiver aggiornato
     */
    Caregiver updateCaregiver(Long id, CaregiverDTO caregiver);

    /**
     * Elimina un caregiver
     * @param caregiver Caregiver da eliminare
     * @return Messaggio di conferma
     */
    String deleteCaregiver(Caregiver caregiver);
    
    Optional<Caregiver> findByFiscalCode(String fiscalCode);
} 

File: front-end/angular/src/app/services/caregiver.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { Caregiver } from '../models/caregiver.model';
import { tap, catchError } from 'rxjs/operators';
import { CaregiverDTO } from '../models/caregiver-dto.model';
import { CaregiverResponse } from '../models/caregiver-response.model';
import { ApiResponse } from '../models/api-response.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CaregiverService {
  private apiUrl = `${environment.apiUrl}/caregivers`;

  constructor(private http: HttpClient) {}

  // Metodo per recuperare i dettagli di un caregiver specifico
  getCaregiverById(caregiverId: number): Observable<any> {
    return this.http.get(`${this.apiUrl}/${caregiverId}`).pipe(
      tap(response => {
        console.log(`Risposta dal server: recuperate info sul caregiver con ID ${caregiverId}:`, response);
      })
    );
  }

  // Metodo per creare un nuovo caregiver
  createCaregiver(caregiverDTO: CaregiverDTO): Observable<CaregiverResponse> {
    return this.http.post<CaregiverResponse>(this.apiUrl, caregiverDTO).pipe(
      tap(response => {
        console.log('Caregiver creato con successo(sto nel service):', response);
      }),
      catchError(this.handleError)
    );
  }

  // Metodo per aggiornare un caregiver esistente
  updateCaregiver(id: number, caregiverDTO: CaregiverDTO): Observable<CaregiverResponse> {
    return this.http.put<CaregiverResponse>(`${this.apiUrl}/${id}`, caregiverDTO).pipe(
      tap(response => {
        console.log(`Caregiver con ID ${id} aggiornato con successo:`, response);
      }),
      catchError(this.handleError)
    );
  }

  private handleError(error: HttpErrorResponse) {
    let errorMessage = 'Si è verificato un errore durante la creazione del caregiver.';
    
    if (error.error instanceof ErrorEvent) {
      // Errore lato client
      errorMessage = `Errore: ${error.error.message}`;
    } else {
      // Errore lato server
      errorMessage = `Il server ha restituito il codice ${error.status}: ${error.error.message || error.message}`;
    }
    
    console.error(errorMessage);
    return throwError(() => new Error(errorMessage));
  }

  // Metodo per cercare un caregiver per codice fiscale
  findByFiscalCode(fiscalCode: string): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/search-by-fiscal-code`, {
      params: { fiscalCode }
    }).pipe(
      tap(response => {
        if (response.success) {
          console.log('Caregiver trovato(questa stampa viene fatta nel CaregiverService):', response.data);
        } else {
          console.log('Caregiver NON trovato(questa stampa viene fatta nel CaregiverService):', response.message);
        }
      }),
      catchError(this.handleError)
    );
  }
} 